Zend/zend_compile.c

```diff
@@ -1257,7 +1257,7 @@ zend_string *zend_type_to_string_resolved(zend_type type, zend_class_entry *scop
 
 	if (type_mask & MAY_BE_NULL) {
 		bool is_union = !str || memchr(ZSTR_VAL(str), '|', ZSTR_LEN(str)) != NULL;
-		if (!is_union) {
+		if (!is_union && !zend_string_equals_literal(str, "false")) {
```

Comment:
From what I understood, ``?T`` is a "legacy" presentation as it usually implies in Reflection it will be a ``ReflectionNamedType`` which for BC reason needs to be maintained for all other nullable types, but as it's a new usage of types having a ``ReflectionUnionType`` be generated makes more sense, and therefore being able to distinguish what Reflection will produce seems sensible to me, but if it's just about the presentation I can change it back.

Reply:
I agree with @nikic on this one, consistency is important. (Also, there is nothing legacy in the current representation of nullable types :shrug: .)

Reply:
Personally I'd remove the special casing both in the rendering here and reflection and always return this like other `?T` types.

FWIW I do think we should switch reflection to always return ReflectionUnionType in the future, but it seems weird to change this only for the one case of `false|null`, no? Or is the some particular reason why this case needs different representation?

Reply:
The main reason is that IMHO a ReflectionUnionType is more correct and because it is new why no use the "proper" reflection case, and then presenting the type as ``false|null`` makes it clear that it will return a ReflectionUnionType and not a ReflectionNamedType.

Moreover, if we do such a switch I'd suppose it would only be possible in PHP 9, which is still some time away.

@nicolas-grekas I wasn't implying one should not be able to use ``?T``, but IMHO ``T|null`` is a better presentation, especially as currently this implies you will not get ReflectionNamedType.

If we do end up using a ReflectionNamedType for ``false|null`` than obviously the presentation should be ``?false``.

Reply:
I think consistency is the most important factor here. There is no reason why this new type should be different.

Reply:
I did inform about the Reflection implementation in the RFC, and that got voted as is, I don't mind changing this. But this feels this at least needs a discussion again on internals just to ask if it is fine to change it to be in line with the rest.